Support for cubic

Aperence opened the pull request:
See #402  for the details.

This PR provide support for cubic, and support for selecting the maximum datagram size.

Exemple of usage : 
```py
from aioquic.quic.congestion.cubic import CubicCongestionControl
from aioquic.quic.congestion.reno import RenoCongestionControl
from aioquic.quic.congestion.slow_starts import HyStart

if (args.cca == "reno"):
    cca = RenoCongestionControl(max_datagram_size=K_MAX_DATAGRAM_SIZE, callback=callback, slow_start=HyStart(max_datagram_size=K_MAX_DATAGRAM_SIZE))
elif (args.cca == "cubic"):
    cca = CubicCongestionControl(max_datagram_size=K_MAX_DATAGRAM_SIZE, callback=callback, reno_friendly_activated=args.reno_friendly)
else:
    exit(1)

conf = QuicConfiguration(alpn_protocols=["quic-testing"], is_client=True, congestion_control=cca)
```

jlaine commented on src/aioquic/quic/configuration.py:
```diff
@@ -1,7 +1,8 @@
 from os import PathLike
 from re import split
-from typing import Any, List, Optional, TextIO, Union
+from typing import Any, List, Optional, TextIO, Union, Dict
```
`Dict` is not used

jlaine commented on src/aioquic/quic/logger.py:
```diff
@@ -32,6 +32,20 @@
+def is_builtin_class(Object):
+    return Object.__class__.__module__ == 'builtins'
+
+def get_dataclass_attr(Object) -> dict:
```
Why is this needed?

Aperence replied:
It was needed for a simpler logging of the BBR/BBRv2 congestion control, where many attributes were present, it was thus better to have some method that automatically gets all its attributes.

This won't be needed for cubic, so I'll simply remove it for now

jlaine commented:
This PR still seems to include unrelated code such as changes to slowstart, minmax.. can we cut it down to the bare minimum to support Cubic please?

Aperence commented:
Hi, 

First, thanks for your feedback.

I have normally removed all content that weren't related to cubic (as slow-start, minmax, get_dataclass_attr, ...).

Only the code for cubic should remain now.

Best,

jlaine commented on src/aioquic/quic/congestion/congestion.py:
```diff
@@ -0,0 +1,155 @@
+        # recovery is a QuicPacketRecovery instance
+        self.recovery = recovery
+
+    def on_init(self, *args, **kwargs):
+        pass
```
This is unused AFAICT?

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,222 @@
+
+    def __init__(self, max_datagram_size : int, callback=None, reno_friendly_activated = True) -> None:
+        super().__init__(max_datagram_size=max_datagram_size, callback=callback)
+        global K_CUBIC_ADDITIVE_INCREASE
```
Uhoh, this doesn't look good, this is de-factor not a constant

jlaine commented on src/aioquic/quic/congestion/congestion.py:
```diff
@@ -0,0 +1,155 @@
+        return False
+    
+
+def Now():
+    return datetime.timestamp(datetime.now())
```
I'm not sure I'm comfortable with this, until now all timestamps have been provided as input to the congestion controller's method. 

If I remember correctly, this timestamp is handed to us by the event loop, and there is no guarantee its aligns with `datetime.now`.

Aperence commented:
Hello, 

I modified the content you pinpointed above, feel free to still give me feedback is anything is still weird or should be changed :D

Have a nice day,

jlaine commented on examples/http3_client.py:
```diff
@@ -545,6 +550,14 @@ async def main(
+        elif args.congestion_control == "reno":
+            configuration.congestion_control = RenoCongestionControl()
+        else:
+            print("Invalid congestion control algorithm")
+            exit(127)
```
```suggestion
            raise Exception("Invalid congestion control algorithm")
```

Aperence commented on src/aioquic/quic/congestion/__init__.py:
```diff
@@ -0,0 +1 @@
+from .congestion import *
```
This feels indeed a better way to do it

jlaine commented:
OK we're getting close to a mergeable state! Would you mind rebasing on top of `main` and squashing down all your commits so we can get a CI run?

Aperence commented:
Normally this should be done, I squashed all the commits into one (4cf275a34e0c8fa30c29b9e9441f0adb31e4015a)

jlaine commented on src/aioquic/quic/congestion/__init__.py:
```diff
@@ -0,0 +1,145 @@
+        if self.callback:
+            self.callback(CongestionEvent.RTT_MEASURED, self)
+
+    def get_congestion_window(self) -> int:
+        # return the cwnd in number of bytes
+        return self.cwnd   
+    
+    def _set_congestion_window(self, value):
+        self.cwnd = value
```
I'm having trouble understanding why we have these accessors?

Aperence replied:
It is used for having an uniform access to the congestion window, as the variable names might change according to the implementation and the developer, thus by providing a method that must be implemented on each class extending QuicCongestionControll, we ensure that whatever the variable name is, we can still access the congestion window (Note however that the setter is unused, so I'll just remove it)

jlaine commented on src/aioquic/quic/congestion/__init__.py:
```diff
@@ -0,0 +1,145 @@
+    def on_packets_lost(self, packets: Iterable[QuicSentPacket], now: float) -> None:
+        if self.callback:
+            self.callback(CongestionEvent.PACKET_LOST, self)
+        if type(self) == QuicCongestionControl:
+            for packet in packets:
+                # don't call this if it is a superclass that runs
+                self.data_in_flight -= packet.sent_bytes
```
This looks very odd..

Aperence replied:
It is an artifact of how bytes in flight was used, I'll refactor this

jlaine commented on src/aioquic/quic/congestion/__init__.py:
```diff
@@ -0,0 +1,145 @@
+    def get_ssthresh(self) -> Optional[int]: 
+        pass
+
+    def get_bytes_in_flight(self) -> int:
+        return self.data_in_flight
```
Same as above, why do we need this accessor, especially since `data_in_flight` doesn't seem to be set anywhere other than at initialisation?

Aperence replied:
Same as above, to provide an uniform interface for getting the bytes in flight, this is needed by `recovery.py` at line 134, which is used typically in `connection.py`, line 580 (this was already there before, just changed it to create a getter)

jlaine commented on src/aioquic/quic/congestion/__init__.py:
```diff
@@ -0,0 +1,145 @@
+    def get_bytes_in_flight(self) -> int:
+        return self.data_in_flight
+
+    def log_callback(self) -> Dict[str, Any]:
```
This doesn't look like it's actually a "callback", it looks more like `get_log_data` or something?

Aperence replied:
It is indeed used to get the logging data for the congestion controller, thus I renamed it

jlaine commented:
I've been reading more of the code and I'm not entirely comfortable with how it's architectured. Could you take a look at PR #432 which *only* moves code around and introduces a base class for congestion control algorithms?

rmarx commented:
Hey both,

(just for the QLOG stuff ;))

qlog is actually intended to be extensible, so implementations can log whatever extra data they care about. We indeed don't want to over-specify for the different CC algorithms, but other implementations have certainly also logged their own fields for that. The one thing you can't count on of course is that general tools (like QVIS) will do anything more with your custom fields than just show them (which is what qvis does). 

Looking at the specific fields here, they seem sensible. My suggestion would be to start the fields with `cubic-` though, and make them lowercase (that IS required by qlog ;)), so `cubic-phase` or `cubic-wmax`.

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,210 @@
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
+        self.now = now
+        super().on_rtt_measurement(latest_rtt, now)
+        # check whether we should exit slow start
+        if self.ssthresh is None and self._rtt_monitor.is_rtt_increasing(
+            latest_rtt, now
+        ):
+            self._cwnd_prior = self.congestion_window
```
This code does nothing, `self._cwnd_prior` is never used anywhere?

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,210 @@
+
+    def on_packet_acked(self, packet: QuicSentPacket, now: float) -> None:
+        super().on_packet_acked(packet, now)
+        rtt = self.recovery._rtt_smoothed
```
No, the congestion control should not have a handle to the `QuicPacketRecovery`. Is it actually the smooth RTT we want here or the latest?

Aperence replied:
It should be according to the RFC the weighted average RTT

Aperence replied:
But I'm wondering if we couldn't use the latest RTT indeed ?

jlaine replied:
I don't know the RFCs enough, if you need it add a `smoothed_rtt` keyword argument to `on_rtt_measurement`

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,210 @@
+    def reset(self):
+        self.congestion_window = K_INITIAL_WINDOW * self._max_datagram_size
+
+        self._cwnd_prior = None
+        self._cwnd_epoch = None
+        self._t_epoch = None
+        self._W_max = None
+        self._first_slow_start = True
+        self._starting_congestion_avoidance = False
+        self.K = 0
+        self._W_est = 0
+        self._cwnd_epoch = 0
+        self._t_epoch = 0
+        self._W_max = self.congestion_window
```
There are multiple redundant initialisations of the same variables..

Aperence replied:
Fixed

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,210 @@
+    def is_concave(self):
+        return self.congestion_window < self._W_max
+    
+    def is_convex(self):
+        return self.congestion_window >= self._W_max
```
This is not used, it's just a negation of `is_concave`

jlaine commented on examples/http3_client.py:
```diff
@@ -518,6 +518,9 @@ async def main(
     parser.add_argument(
         "--zero-rtt", action="store_true", help="try to send requests using 0-RTT"
     )
+    parser.add_argument(
+        "--congestion-control", type=str, help="which congestion control algorithm to use (reno, cubic)"
+    )
```
Not needed, already in `main`

jlaine commented on examples/http3_server.py:
```diff
@@ -564,6 +564,9 @@ async def main(
     parser.add_argument(
         "-v", "--verbose", action="store_true", help="increase logging verbosity"
     )
+    parser.add_argument(
+        "--congestion-control", type=str, help="which congestion control algorithm to use (reno, cubic)"
+    )
```
Not needed, already in `main`

jlaine commented on src/aioquic/quic/congestion/base.py:
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
     ssthresh: Optional[int] = None
 
-    def __init__(self, *, max_datagram_size: int) -> None:
+    def __init__(self, *, max_datagram_size: int, recovery) -> None:
         self.congestion_window = K_INITIAL_WINDOW * max_datagram_size
+        self.recovery = recovery
```
No, let's not have a reference loop please

Aperence replied:
How should I get the smoothed RTT in another way ?

Aperence replied:
Changed to latest RTT instead

jlaine commented on src/aioquic/quic/congestion/base.py:
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
 
     @abc.abstractmethod
-    def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_acked(self, *, now : float, packet: QuicSentPacket) -> None:
```
Why do you need `now`, you already have `packet.sent_time`?

jlaine commented on src/aioquic/quic/congestion/base.py:
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
 
     @abc.abstractmethod
-    def on_packet_sent(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_sent(self, *, now : float, packet: QuicSentPacket) -> None:
```
Why do you need `now`, you already have `packet.sent_time`?

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,204 @@
+
+        self.now = 0
+
+    def better_cube_root(self, x):
+        if (x < 0):
+            # avoid precision errors that make the cube root returns an imaginary number
+            return -((-x)**(1./3.))
+        else:
+            return (x)**(1./3.)
```
Move this to a standalone function, it will avoid re-defining it in `test_cubic.py`

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -62,7 +63,10 @@ def update_bucket(self, now: float) -> None:
 
     def update_rate(self, congestion_window: int, smoothed_rtt: float) -> None:
-        pacing_rate = congestion_window / max(smoothed_rtt, K_MICRO_SECOND)
+        if self.pacing_rate == None or self.pacing_rate == 0:
+            pacing_rate = congestion_window / max(smoothed_rtt, K_MICRO_SECOND)
+        else:
+            pacing_rate = self.pacing_rate
```
Why are we changing this?

Aperence replied:
Woopsie, seems like an artifact of BBR (where the congestion controller needed to set the pacing rate), just removed it

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -115,17 +119,17 @@ def __init__(
     @property
     def bytes_in_flight(self) -> int:
-        return self._cc.bytes_in_flight
+        return int(self._cc.bytes_in_flight)
```
This is already an int, no need to cast it.

jlaine replied:
> In cubic, this can be a float, leading to some unexpected error, so I just casted it when trying to access it

Uh no I don't see how this can be a float, we start with 0 and all we do is add / subtract packet lengths. This cannot and must not be a float.

jlaine commented on src/aioquic/quic/congestion/base.py:
```diff
@@ -1,5 +1,5 @@
 import abc
-from typing import Dict, Iterable, Optional, Protocol
+from typing import Dict, Iterable, Optional, Protocol, Any
```
Please run the linters (black / ruff / mypy) locally to avoid CI runs,  this is guaranteed to not be OK as it's not sorted

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,206 @@
+            self._starting_congestion_avoidance = True  # restart a new congestion avoidance phase
+
+
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
```
`rtt` not `latest_rtt`

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,206 @@
+
+        self.last_ack = None
+
+        self.now = 0
```
Please remove this, the controller cannot know the "current time" as it has no clock. From what I can see the only place you use `self.now` is in `get_log_data`, you can pass it in there.

jlaine commented:
Please revert your changes to `reno.py`, the only change should be the addition of `now` in `on_packet_acked`.

jlaine commented on src/aioquic/quic/connection.py:
```diff
@@ -3318,4 +3319,4 @@ def _write_streams_blocked_frame(
                     limit=limit,
                 )
-            )
+            )
```
Please revert all changes to `connection.py`, they are not needed

jlaine commented on src/aioquic/quic/logger.py:
```diff
@@ -32,7 +32,6 @@
 def hexdump(data: bytes) -> str:
     return binascii.hexlify(data).decode("ascii")
 
-
```
Please revert all changes to `logger.py`, they are not needed

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -1,8 +1,9 @@
 import logging
 import math
 from typing import Any, Callable, Dict, Iterable, List, Optional
+from datetime import datetime
```
This is sure to fail linting, use `black` and `ruff --fix` locally to fix everything in `src`

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -238,7 +238,7 @@ def on_ack_received(
 
             # inform congestion controller
-            self._cc.on_rtt_measurement(now=now, rtt=latest_rtt)
+            self._cc.on_rtt_measurement(now=now, latest_rtt=latest_rtt)
```
Revert this

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -263,7 +263,7 @@ def on_loss_detection_timeout(self, *, now: float) -> None:
             self.reschedule_data(now=now)
 
-    def on_packet_sent(self, *, packet: QuicSentPacket, space: QuicPacketSpace) -> None:
+    def on_packet_sent(self, packet: QuicSentPacket, space: QuicPacketSpace) -> None:
```
Revert this

jlaine commented on tests/test_recovery.py:
```diff
@@ -1,7 +1,20 @@
 from unittest import TestCase
 
+<<<<<<< HEAD
```
Revert all these changes

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,206 @@
+K_CUBIC_LOSS_REDUCTION_FACTOR = 0.7
+K_CUBIC_MAX_IDLE_TIME = 2   # reset the cwnd after 2 seconds of inactivity
+
+def better_cube_root(self, x):
```
self? While we are at it, please add type annotations otherwise `mypy` will fail

jlaine commented on src/aioquic/quic/congestion/reno.py:
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
                 self.congestion_window += count * self._max_datagram_size
 
-    def on_packet_sent(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_sent(self, packet: QuicSentPacket) -> None:
```
The switch to keyword-only arguments was on purpose, revert this

jlaine commented on src/aioquic/quic/congestion/reno.py:
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
         self.bytes_in_flight += packet.sent_bytes
 
-    def on_packets_expired(self, *, packets: Iterable[QuicSentPacket]) -> None:
+    def on_packets_expired(self, packets: Iterable[QuicSentPacket]) -> None:
```
Ditto

jlaine commented on src/aioquic/quic/congestion/reno.py:
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
             self.bytes_in_flight -= packet.sent_bytes
 
-    def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> None:
+    def on_packets_lost(self, packets: Iterable[QuicSentPacket], now: float) -> None:
```
Ditto

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,206 @@
+from typing import Iterable, Dict, Any
+
+# cubic specific variables (see https://www.rfc-editor.org/rfc/rfc9438.html#name-definitions)
+K_CUBIC_K = 1    
```
Where is `K_CUBIC_K` used?

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,206 @@
+
+            t = now - self._t_epoch
+            
+            target = None
+            if (self.W_cubic(t + self.rtt) < self.congestion_window):
+                target = self.congestion_window
+            elif (self.W_cubic(t + self.rtt) > 1.5*self.congestion_window):
+                target = self.congestion_window*1.5
+            else:
+                target = self.W_cubic(t + self.rtt)
```
No need to re-calculate this 3 times, just assign `target = self.W_cubic(t + self.rtt)` then clamp it

jlaine commented on src/aioquic/quic/congestion/reno.py:
```diff
@@ -66,12 +66,22 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
 
-    def on_rtt_measurement(self, *, now: float, rtt: float) -> None:
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
         # check whether we should exit slow start
         if self.ssthresh is None and self._rtt_monitor.is_rtt_increasing(
-            now=now, rtt=rtt
+            rtt=latest_rtt, now=now
```
Revert this too

jlaine commented on src/aioquic/quic/congestion/reno.py:
```diff
@@ -66,12 +66,22 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
+        data = {
+            "cwnd" : self.congestion_window
+        }
+        if (self.ssthresh == None):
```
```suggestion
        if self.ssthresh is None:
```

Aperence commented:
Ok, I've reverted normally every changed except for cubic, and logs, I've also run the lint checks locally (ruff + black + mypy, thank you btw because I didn't know about these tools at all) and I fixed normally everything that was listed locally

Aperence commented:
My black version locally seem however to differ compared to the one used on CI, as it fails here, I can fix this (removing a blank line), if needed, but I'm waiting for you to confirm as I might have abused you CI instances with all these push, my apologies for this :cry:

jlaine commented:
It looks as though you still have one linter error left :)

The code is also going to need thorough tests to check everything is working in all scenarios. I would suggest you try writing a `test_recovery_cubic.py` based on `test_recovery_reno.py`. Full test coverage is a requirement to merge :)

jlaine commented on src/aioquic/quic/congestion/base.py:
```diff
@@ -40,6 +40,12 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
         ...  # pragma: no cover
 
+    def get_log_data(self) -> Dict[str, Any]:
+        data = {"cwnd": self.congestion_window}
```
Looks like you dropped `bytes_in_flight` here

jlaine commented on src/aioquic/quic/recovery.py:
```diff
@@ -334,10 +334,7 @@ def _get_loss_space(self) -> Optional[QuicPacketSpace]:
-            "cwnd": self._cc.congestion_window,
-        }
+        data: Dict[str, Any] = self._cc.get_log_data()
         if self._cc.ssthresh is not None:
```
This doesn't look needed, we're already reporting `ssthresh` in `QuicCongestionControl.get_log_data()`

jlaine commented on src/aioquic/quic/congestion/cubic.py:
```diff
@@ -0,0 +1,216 @@
+
+        self._first_slow_start = True
+        self._starting_congestion_avoidance = False
+        self.K: float = 0.0
```
If you initialise K with a float you don't need the type definition

Aperence commented:
Normally, I've fixed these last issues + added a test for cubic based on `test_recovery_cubic.py`. I'll try to check later the code that wasn't covered by thos tests

jlaine commented:
> Normally, I've fixed these last issues + added a test for cubic based on `test_recovery_cubic.py`. I'll try to check later the code that wasn't covered by thos tests

Sure! You can get coverage reports locally too:

```shell
pip install coverage
coverage run -m unittest tests/test_recovery_cubic.py
coverage html
firefox htmlcov/index.html
```

jlaine commented:
Wow, amazing work, thanks so much for your patience! Is the `test_cubic.py` still useful or is it superseded by your `test_recovery_cubic.py`? If it's still useful maybe merge it into `test_recovery_cubic.py`?

Disclaimer: I am not able to determine whether the code works as intended, or follows specs. Would you mind taking the code in its current form for a spin and checking it still behaves as you want it to?

Once this is done I'd be happy to merge the code!

Aperence commented:
Hi, 

First, thank you for all your feedback during this PR, I learned quite a lot about CI tools to ensure some code quality !
Secondly, about the `test_cubic.py`, it is still useful, so I'll just merge it in `test_recovery_cubic.py` as you recommended, this should centralize all the tests for the cubic congestion controller.

The implementation seems to still work fine, here are some examples on how it performs:

With no delay on localhost:
![image](https://github.com/aiortc/aioquic/assets/78789735/b4edab1b-bd5f-4c0b-b149-7db28b8f3fde)

With a delay of 100ms on localhost:
![image](https://github.com/aiortc/aioquic/assets/78789735/538b6b1d-03be-4e70-87e3-c56a84c22ded)

We can clearly still see the cubic growth when the delay is quite important (second plot), and the reno friendly region when we have a reasonable delay (first plot), thus I think we can affirm that it still work as intended

jlaine commented:
Alright, thanks for the checks! Are these graphs produced from the QLOG output? I saw an article which mentioned you had set up a test bench as part of your University work?

If wouldn't mind squashing your commits down to a single commit with a clean commit message I'd be happy to hit the "merge" button.

Aperence commented:
Yup, these graphs are generated based on the QLOG files, the script can be found [here](https://github.com/Aperence/linfo1341-congestion-student/blob/main/src/plot_congestion.py) if you are interested with it (it support normally any congestion control, as it only check the presence of the cwnd field in the logs), but some features (W_max, ...) might not be up to date with the latest modifications for the logs (but the core for plotting losses and cwnd does). You can also select a particular field if you want to plot other things (for example ssthresh, bytes_in_flight, ...)

Normally, the commits should be squashed also !

jlaine commented:
Merged, thanks a lot! 🎉

I hope we will have the pleasure of seing more of your work :)
